### PR TITLE
Markdown output from bench-compare

### PR DESCRIPTION
**Pull request by kwxm**
I've added a new script `bench-compare-markdown` which produces a markdown table of benchmark comparison results.  I've tired to make it so that the raw output's pretty readable on its own, but it might look a bit strange if some of the output is unusually wide (but the rendered output should still be OK in that case).  I'd like to keep the original version because it's probably better for pasting into Slack or converting into Latex, but maybe that doesn't need to be in the repository.  Both scripts now take an extra two arguments at the end which are used for column headers, so you can stick dates or commit hashes in there.  The headers default to `Old` and `New`.  

The original script was just a throwaway that I produced to easily compare two lots of Criterion output and I implemented it in Awk because that made it easier to discard all the irrelevant stuff that Cirterion produces (plus build messages from cabal and so on).  We _could_ get just the essential information from Criterion using a `--csv`/`--junit`/`--json` option, and that might be easier to process using Haskell if you want to be purist.  It'd be a bit clumsier on the command line but that wouldn't be a problem for automated use.

Examples of output in comments below.

Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Comment by kwxm** [comment]
Raw output:
```
| Script                                   | 15th June  | 30th June  | Change   |
| :----                                    | :-----:    | :-----:    | :-----:  |
| crowdfunding/crowdfunding-success-1      |  696.4 μs  |  653.1 μs  |   -6.2%  |
| crowdfunding/crowdfunding-success-2      |  698.1 μs  |  655.2 μs  |   -6.1%  |
| crowdfunding/crowdfunding-success-3      |  696.9 μs  |  654.5 μs  |   -6.1%  |
| currency/currency-1                      |  630.8 μs  |  583.2 μs  |   -7.5%  |
| currency/currency-2                      |  814.8 μs  |  795.9 μs  |   -2.3%  |
| escrow/escrow-redeem_1-1                 |  971.3 μs  |  963.9 μs  |   -0.8%  |
| escrow/escrow-redeem_1-2                 |  970.9 μs  |  965.7 μs  |   -0.5%  |
| escrow/escrow-redeem_2-1                 |  1.115 ms  |  1.106 ms  |   -0.8%  |
| escrow/escrow-redeem_2-2                 |  1.113 ms  |  1.106 ms  |   -0.6%  |
| escrow/escrow-redeem_2-3                 |  1.114 ms  |  1.107 ms  |   -0.6%  |
| escrow/escrow-refund-1                   |  482.7 μs  |  458.6 μs  |   -5.0%  |
| future/future-increase-margin-1          |  629.1 μs  |  586.4 μs  |   -6.8%  |
| future/future-increase-margin-2          |  811.6 μs  |  799.8 μs  |   -1.5%  |
| future/future-increase-margin-3          |  1.415 ms  |  1.398 ms  |   -1.2%  |
| future/future-increase-margin-4          |  1.417 ms  |  1.398 ms  |   -1.3%  |
| future/future-increase-margin-5          |  1.308 ms  |  1.296 ms  |   -0.9%  |
| future/future-increase-margin-6          |  1.935 ms  |  1.805 ms  |   -6.7%  |
| future/future-pay-out-1                  |  631.7 μs  |  585.0 μs  |   -7.4%  |
| future/future-pay-out-2                  |  814.9 μs  |  797.3 μs  |   -2.2%  |
| future/future-pay-out-3                  |  1.422 ms  |  1.394 ms  |   -2.0%  |
| future/future-pay-out-4                  |  1.424 ms  |  1.393 ms  |   -2.2%  |
| future/future-pay-out-5                  |  1.827 ms  |  1.801 ms  |   -1.4%  |
| future/future-settle-early-1             |  631.4 μs  |  582.9 μs  |   -7.7%  |
| future/future-settle-early-2             |  813.8 μs  |  797.1 μs  |   -2.1%  |
| future/future-settle-early-3             |  1.418 ms  |  1.389 ms  |   -2.0%  |
| future/future-settle-early-4             |  1.423 ms  |  1.399 ms  |   -1.7%  |
| future/future-settle-early-5             |  1.442 ms  |  1.428 ms  |   -1.0%  |
| game-sm/game-sm-success-1                |  1.010 ms  |  1.002 ms  |   -0.8%  |
| game-sm/game-sm-success-2                |  539.3 μs  |  603.5 μs  |  +11.9%  |
| game-sm/game-sm-success-3                |  1.451 ms  |  1.453 ms  |   +0.1%  |
| game-sm/game-sm-success-4                |  752.9 μs  |  717.9 μs  |   -4.6%  |
| game-sm/game-sm-success_2-1              |  1.009 ms  |  1.002 ms  |   -0.7%  |
| game-sm/game-sm-success_2-2              |  539.3 μs  |  602.6 μs  |  +11.7%  |
| game-sm/game-sm-success_2-3              |  1.447 ms  |  1.447 ms  |    0.0%  |
| game-sm/game-sm-success_2-4              |  751.8 μs  |  719.9 μs  |   -4.2%  |
| game-sm/game-sm-success_2-5              |  1.477 ms  |  1.427 ms  |   -3.4%  |
| game-sm/game-sm-success_2-6              |  754.7 μs  |  717.1 μs  |   -5.0%  |
| multisig-sm/multisig-sm-01               |  1.073 ms  |  1.061 ms  |   -1.1%  |
| multisig-sm/multisig-sm-02               |  1.082 ms  |  1.066 ms  |   -1.5%  |
| multisig-sm/multisig-sm-03               |  1.100 ms  |  1.080 ms  |   -1.8%  |
| multisig-sm/multisig-sm-04               |  1.115 ms  |  1.094 ms  |   -1.9%  |
| multisig-sm/multisig-sm-05               |  1.312 ms  |  1.290 ms  |   -1.7%  |
| multisig-sm/multisig-sm-06               |  1.170 ms  |  1.059 ms  |   -9.5%  |
| multisig-sm/multisig-sm-07               |  1.082 ms  |  1.061 ms  |   -1.9%  |
| multisig-sm/multisig-sm-08               |  1.101 ms  |  1.165 ms  |   +5.8%  |
| multisig-sm/multisig-sm-09               |  1.113 ms  |  1.093 ms  |   -1.8%  |
| multisig-sm/multisig-sm-10               |  1.306 ms  |  1.298 ms  |   -0.6%  |
| ping-pong/ping-pong-1                    |  833.4 μs  |  803.7 μs  |   -3.6%  |
| ping-pong/ping-pong-2                    |  834.2 μs  |  803.5 μs  |   -3.7%  |
| ping-pong/ping-pong_2-1                  |  530.5 μs  |  496.2 μs  |   -6.5%  |
| prism/prism-1                            |  498.8 μs  |  472.5 μs  |   -5.3%  |
| prism/prism-2                            |  1.096 ms  |  1.070 ms  |   -2.4%  |
| prism/prism-3                            |  987.0 μs  |  981.6 μs  |   -0.5%  |
| pubkey/pubkey-1                          |  439.7 μs  |  415.0 μs  |   -5.6%  |
| stablecoin/stablecoin_1-1                |  2.095 ms  |  2.088 ms  |   -0.3%  |
| stablecoin/stablecoin_1-2                |  633.0 μs  |  589.9 μs  |   -6.8%  |
| stablecoin/stablecoin_1-3                |  2.363 ms  |  2.353 ms  |   -0.4%  |
| stablecoin/stablecoin_1-4                |  690.1 μs  |  641.4 μs  |   -7.1%  |
| stablecoin/stablecoin_1-5                |  2.884 ms  |  2.984 ms  |   +3.5%  |
| stablecoin/stablecoin_1-6                |  723.8 μs  |  824.7 μs  |  +13.9%  |
| stablecoin/stablecoin_2-1                |  2.101 ms  |  2.082 ms  |   -0.9%  |
| stablecoin/stablecoin_2-2                |  635.7 μs  |  589.4 μs  |   -7.3%  |
| stablecoin/stablecoin_2-3                |  2.376 ms  |  2.355 ms  |   -0.9%  |
| stablecoin/stablecoin_2-4                |  689.5 μs  |  644.8 μs  |   -6.5%  |
| token-account/token-account-1            |  594.3 μs  |  556.8 μs  |   -6.3%  |
| token-account/token-account-2            |  697.9 μs  |  653.8 μs  |   -6.3%  |
| token-account/token-account-3            |  887.2 μs  |  884.8 μs  |   -0.3%  |
| uniswap/uniswap-1                        |  696.6 μs  |  648.8 μs  |   -6.9%  |
| uniswap/uniswap-2                        |  1.114 ms  |  1.100 ms  |   -1.3%  |
| uniswap/uniswap-3                        |  757.2 μs  |  718.2 μs  |   -5.2%  |
| uniswap/uniswap-4                        |  832.0 μs  |  818.8 μs  |   -1.6%  |
| uniswap/uniswap-5                        |  3.896 ms  |  3.878 ms  |   -0.5%  |
| uniswap/uniswap-6                        |  1.125 ms  |  1.105 ms  |   -1.8%  |
| uniswap/uniswap-7                        |  2.826 ms  |  2.803 ms  |   -0.8%  |
| uniswap/uniswap-8                        |  1.075 ms  |  1.052 ms  |   -2.1%  |
| vesting/vesting-1                        |  904.7 μs  |  897.3 μs  |   -0.8%  |
| marlowe/trustfund/trustfund-1            |  2.125 ms  |  2.072 ms  |   -2.5%  |
| marlowe/trustfund/trustfund-2            |  1.520 ms  |  1.517 ms  |   -0.2%  |
| marlowe/zerocoupon/zerocoupon-1          |  2.196 ms  |  2.168 ms  |   -1.3%  |
| marlowe/zerocoupon/zerocoupon-2          |  1.316 ms  |  1.319 ms  |   +0.2%  |
```

**Comment by kwxm** [comment]
Rendered output:

| Script                                   | 15th June  | 30th June  | Change   |
| :----                                    | :-----:    | :-----:    | :-----:  |
| crowdfunding/crowdfunding-success-1      |  696.4 μs  |  653.1 μs  |   -6.2%  |
| crowdfunding/crowdfunding-success-2      |  698.1 μs  |  655.2 μs  |   -6.1%  |
| crowdfunding/crowdfunding-success-3      |  696.9 μs  |  654.5 μs  |   -6.1%  |
| currency/currency-1                      |  630.8 μs  |  583.2 μs  |   -7.5%  |
| currency/currency-2                      |  814.8 μs  |  795.9 μs  |   -2.3%  |
| escrow/escrow-redeem_1-1                 |  971.3 μs  |  963.9 μs  |   -0.8%  |
| escrow/escrow-redeem_1-2                 |  970.9 μs  |  965.7 μs  |   -0.5%  |
| escrow/escrow-redeem_2-1                 |  1.115 ms  |  1.106 ms  |   -0.8%  |
| escrow/escrow-redeem_2-2                 |  1.113 ms  |  1.106 ms  |   -0.6%  |
| escrow/escrow-redeem_2-3                 |  1.114 ms  |  1.107 ms  |   -0.6%  |
| escrow/escrow-refund-1                   |  482.7 μs  |  458.6 μs  |   -5.0%  |
| future/future-increase-margin-1          |  629.1 μs  |  586.4 μs  |   -6.8%  |
| future/future-increase-margin-2          |  811.6 μs  |  799.8 μs  |   -1.5%  |
| future/future-increase-margin-3          |  1.415 ms  |  1.398 ms  |   -1.2%  |
| future/future-increase-margin-4          |  1.417 ms  |  1.398 ms  |   -1.3%  |
| future/future-increase-margin-5          |  1.308 ms  |  1.296 ms  |   -0.9%  |
| future/future-increase-margin-6          |  1.935 ms  |  1.805 ms  |   -6.7%  |
| future/future-pay-out-1                  |  631.7 μs  |  585.0 μs  |   -7.4%  |
| future/future-pay-out-2                  |  814.9 μs  |  797.3 μs  |   -2.2%  |
| future/future-pay-out-3                  |  1.422 ms  |  1.394 ms  |   -2.0%  |
| future/future-pay-out-4                  |  1.424 ms  |  1.393 ms  |   -2.2%  |
| future/future-pay-out-5                  |  1.827 ms  |  1.801 ms  |   -1.4%  |
| future/future-settle-early-1             |  631.4 μs  |  582.9 μs  |   -7.7%  |
| future/future-settle-early-2             |  813.8 μs  |  797.1 μs  |   -2.1%  |
| future/future-settle-early-3             |  1.418 ms  |  1.389 ms  |   -2.0%  |
| future/future-settle-early-4             |  1.423 ms  |  1.399 ms  |   -1.7%  |
| future/future-settle-early-5             |  1.442 ms  |  1.428 ms  |   -1.0%  |
| game-sm/game-sm-success-1                |  1.010 ms  |  1.002 ms  |   -0.8%  |
| game-sm/game-sm-success-2                |  539.3 μs  |  603.5 μs  |  +11.9%  |
| game-sm/game-sm-success-3                |  1.451 ms  |  1.453 ms  |   +0.1%  |
| game-sm/game-sm-success-4                |  752.9 μs  |  717.9 μs  |   -4.6%  |
| game-sm/game-sm-success_2-1              |  1.009 ms  |  1.002 ms  |   -0.7%  |
| game-sm/game-sm-success_2-2              |  539.3 μs  |  602.6 μs  |  +11.7%  |
| game-sm/game-sm-success_2-3              |  1.447 ms  |  1.447 ms  |    0.0%  |
| game-sm/game-sm-success_2-4              |  751.8 μs  |  719.9 μs  |   -4.2%  |
| game-sm/game-sm-success_2-5              |  1.477 ms  |  1.427 ms  |   -3.4%  |
| game-sm/game-sm-success_2-6              |  754.7 μs  |  717.1 μs  |   -5.0%  |
| multisig-sm/multisig-sm-01               |  1.073 ms  |  1.061 ms  |   -1.1%  |
| multisig-sm/multisig-sm-02               |  1.082 ms  |  1.066 ms  |   -1.5%  |
| multisig-sm/multisig-sm-03               |  1.100 ms  |  1.080 ms  |   -1.8%  |
| multisig-sm/multisig-sm-04               |  1.115 ms  |  1.094 ms  |   -1.9%  |
| multisig-sm/multisig-sm-05               |  1.312 ms  |  1.290 ms  |   -1.7%  |
| multisig-sm/multisig-sm-06               |  1.170 ms  |  1.059 ms  |   -9.5%  |
| multisig-sm/multisig-sm-07               |  1.082 ms  |  1.061 ms  |   -1.9%  |
| multisig-sm/multisig-sm-08               |  1.101 ms  |  1.165 ms  |   +5.8%  |
| multisig-sm/multisig-sm-09               |  1.113 ms  |  1.093 ms  |   -1.8%  |
| multisig-sm/multisig-sm-10               |  1.306 ms  |  1.298 ms  |   -0.6%  |
| ping-pong/ping-pong-1                    |  833.4 μs  |  803.7 μs  |   -3.6%  |
| ping-pong/ping-pong-2                    |  834.2 μs  |  803.5 μs  |   -3.7%  |
| ping-pong/ping-pong_2-1                  |  530.5 μs  |  496.2 μs  |   -6.5%  |
| prism/prism-1                            |  498.8 μs  |  472.5 μs  |   -5.3%  |
| prism/prism-2                            |  1.096 ms  |  1.070 ms  |   -2.4%  |
| prism/prism-3                            |  987.0 μs  |  981.6 μs  |   -0.5%  |
| pubkey/pubkey-1                          |  439.7 μs  |  415.0 μs  |   -5.6%  |
| stablecoin/stablecoin_1-1                |  2.095 ms  |  2.088 ms  |   -0.3%  |
| stablecoin/stablecoin_1-2                |  633.0 μs  |  589.9 μs  |   -6.8%  |
| stablecoin/stablecoin_1-3                |  2.363 ms  |  2.353 ms  |   -0.4%  |
| stablecoin/stablecoin_1-4                |  690.1 μs  |  641.4 μs  |   -7.1%  |
| stablecoin/stablecoin_1-5                |  2.884 ms  |  2.984 ms  |   +3.5%  |
| stablecoin/stablecoin_1-6                |  723.8 μs  |  824.7 μs  |  +13.9%  |
| stablecoin/stablecoin_2-1                |  2.101 ms  |  2.082 ms  |   -0.9%  |
| stablecoin/stablecoin_2-2                |  635.7 μs  |  589.4 μs  |   -7.3%  |
| stablecoin/stablecoin_2-3                |  2.376 ms  |  2.355 ms  |   -0.9%  |
| stablecoin/stablecoin_2-4                |  689.5 μs  |  644.8 μs  |   -6.5%  |
| token-account/token-account-1            |  594.3 μs  |  556.8 μs  |   -6.3%  |
| token-account/token-account-2            |  697.9 μs  |  653.8 μs  |   -6.3%  |
| token-account/token-account-3            |  887.2 μs  |  884.8 μs  |   -0.3%  |
| uniswap/uniswap-1                        |  696.6 μs  |  648.8 μs  |   -6.9%  |
| uniswap/uniswap-2                        |  1.114 ms  |  1.100 ms  |   -1.3%  |
| uniswap/uniswap-3                        |  757.2 μs  |  718.2 μs  |   -5.2%  |
| uniswap/uniswap-4                        |  832.0 μs  |  818.8 μs  |   -1.6%  |
| uniswap/uniswap-5                        |  3.896 ms  |  3.878 ms  |   -0.5%  |
| uniswap/uniswap-6                        |  1.125 ms  |  1.105 ms  |   -1.8%  |
| uniswap/uniswap-7                        |  2.826 ms  |  2.803 ms  |   -0.8%  |
| uniswap/uniswap-8                        |  1.075 ms  |  1.052 ms  |   -2.1%  |
| vesting/vesting-1                        |  904.7 μs  |  897.3 μs  |   -0.8%  |
| marlowe/trustfund/trustfund-1            |  2.125 ms  |  2.072 ms  |   -2.5%  |
| marlowe/trustfund/trustfund-2            |  1.520 ms  |  1.517 ms  |   -0.2%  |
| marlowe/zerocoupon/zerocoupon-1          |  2.196 ms  |  2.168 ms  |   -1.3%  |
| marlowe/zerocoupon/zerocoupon-2          |  1.316 ms  |  1.319 ms  |   +0.2%  |

**Comment by kwxm** [comment]
Personally I think the textual output's a bit easier to read because it's more compact.

**Comment by kwxm** [comment]
@michaelpj These tables contain figures from when I switched to the new validation scripts a couple of weeks ago and from yesterday's `master`.  They're kind of interesting.

**Comment by michaelpj** [comment]
I like the tables :)

Do we need to have two scripts? Can't we just have the markdown one? The plain-text version of the markdown one isn't much harder to read than the current one, so I think it still functions okay in that mode

**Comment by kwxm** [comment]
> I like the tables :)
> 
> Do we need to have two scripts? Can't we just have the markdown one? The plain-text version of the markdown one isn't much harder to read than the current one, so I think it still functions okay in that mode

I think the original one's probably better for Slack and stuff (and once or twice I've modified it to compare other things, like the cost model predictions), but there's nothing to stop me from using my own personal copy if you want to remove it from the repository.

**Comment by michaelpj** [comment]
I won't fight about it, but the only thing worse than maintaining a bunch of awk is maintaining two bunches of awk which are nearly but not quite the same...

**Comment by kwxm** [comment]
> I won't fight about it, but the only thing worse than maintaining a bunch of awk is maintaining two bunches of awk which are nearly but not quite the same...

I'll just remove it and keep a secret copy for personal use.

**Comment by michaelpj** [comment]
If we're only going to have one, can we just call it `bench-compare`?

**Comment by michaelpj** [comment]
And e.g. the benchmark job will be broken if we don't have that script, so either we need to give it the new name or change that immediately.

**Comment by kwxm** [comment]
> And e.g. the benchmark job will be broken if we don't have that script, so either we need to give it the new name or change that immediately.

Well I can change that right now, and put the hashes in the results too!  Let's see if that worked...

File: scripts/ci-plutus-benchmark.sh [hunk 1]
```diff
@@ -47,7 +47,7 @@ cabal bench plutus-benchmark:validation >bench-base.log 2>&1
 
 echo "[ci-plutus-benchmark]: Comparing results ..."
 echo -e "Comparing benchmark results of '$BASE_BRANCH_REF' (base) and '$PR_BRANCH_REF' (PR)\n" >bench-compare-result.log
-./plutus-benchmark/bench-compare bench-base.log bench-PR.log >>bench-compare-result.log
+./plutus-benchmark/bench-compare-markdown bench-base.log bench-PR.log "$BASE_BRANCH_REF" "$PR_BRANCH_REF" >>bench-compare-result.log
```
Comment (kwxm) [inline scripts/ci-plutus-benchmark.sh:50]: Should we make the script non-executable and invoke it with `bash` instead?  Maybe that would rely on having a specific shell on the benchmarking machine though.

**Comment by kwxm** [comment]
Damn.  It doesn't work on this branch because it does the comparison in the base commit and `bench-compare-markdown` doesn't exist there.  Oh well, I'm just going to have to have two different files called `bench-compare`.

**Comment by michaelpj** [comment]
>  If we don't, it won't be possible to test any future changes to the script in their own PR.

This is a little tricky. The rest of the pipeline definition will come from the base branch, so there's some risk that e.g. if you changed how arguments were passed tin then you'd get a mismatch with the calling code being from the base branch and `bench-compare` being from the PR branch. So I don't know which is worse. Opinions @gilligan ?

**Comment by kwxm** [comment]
> > If we don't, it won't be possible to test any future changes to the script in their own PR.
> 
> This is a little tricky. The rest of the pipeline definition will come from the base branch, so there's some risk that e.g. if you changed how arguments were passed tin then you'd get a mismatch with the calling code being from the base branch and `bench-compare` being from the PR branch. So I don't know which is worse. Opinions @gilligan ?

I'm going to try switching back to the PR branch at the end, but I agree we have to be careful and need to think about this.

**Comment by gilligan** [comment]
Uhm yeah so I think the fact that the comparison works on 2 separate branches just inherently introduces problems when the "interface" changes. I don't think it makes much sense to try to work around that.

So i'd say try to test locally what you can and then unfortunately the real testing will have to happen after it's merged. FWIW I had to actually work and test all of this on a plutus fork because the GitHub comment event would only trigger from code in master..

**Comment by michaelpj** [comment]
Oh, you were trying to make it go to the PR branch but you got it to check out the base branch...

**Comment by michaelpj** [comment]
Well we can tweak it still.

**Comment by kwxm** [comment]
> Oh, you were trying to make it go to the PR branch but you got it to check out the base branch...

Sorry, I should have said not to merge this!  I was messing about with the benchmarking script and stupidly typed in the wrong branch, so it probably won't work any more.  That's not fatal though.

**Comment by michaelpj** [comment]
I mean, it will work, it just won't use the new script :)

**Pull request merged**